libs/utils/android/workload.py

```diff
@@ -30,113 +30,109 @@ class Workload(object):
-                    class_name = handler.__name__
-                    module_name = module.__name__
-                    # Check if a package is required and is available on target
-                    aw = Workload._AW(module_name, module, class_name, handler)
-                    if (Workload._is_available(target, aw)):
-                        # Keep track of classes which are 'Android.Workload'
-                        Workload._availables[class_name] = aw
-
-        return Workload._availables.keys()
-
-    @staticmethod
-    def _is_available(target, aw):
-        try:
-            package = getattr(aw.ctor, 'package')
-        except AttributeError:
-            # Assume workloads not requiring a package
-            # are always available
-            return True
-
-        # Check for the package being available
-        count = target.execute('pm list packages | grep {} | wc -l'\
-                               .format(package))
-        if int(count) >= 1:
-            return True
-
-        log = logging.getLogger('Workload')
-        log.warning('Package [%s] not installed', package)
-        log.warning('Workload [%s] disabled', aw.class_name)
-        return False
 
     def __init__(self, test_env):
         """
         Initialized workloads available on the specified test environment
 
         test_env: target test environmen
         """
-        self.te = test_env
-        self.target = test_env.target
+        self._te = test_env
+        self._target = test_env.target
         self._log = logging.getLogger('Workload')
 
-        wloads = Workload.availables(self.target)
-        self._log.info('Workloads available on target:')
-        self._log.info('  %s', wloads)
+        # Set of data reported in output of each run
+        self.trace_file = None
+        self.nrg_report = None
+
+        self._packages = None
+        self._availables = {}
+
+        # Build list of supported workloads available in the target
+        self._check_availables()
 
     def _adb(self, cmd):
-        return 'adb -s {} {}'.format(self.target.adb_name, cmd)
+        return 'adb -s {} {}'.format(self._target.adb_name, cmd)
+
+    def _check_availables(self):
+        """
+        List the supported android workloads which are available on the target
+        """
+        # Getting the list of installed packages
+        self._packages = self._target.execute('pm list packages').splitlines()
+        self._log.debug('Packages:\n%s', self._packages)
+
+        self._log.debug('Building list of available workloads...')
+        for sc in Workload.__subclasses__():
+            self._log.debug('Checking workload [%s]...', sc.__name__)
+            if 'package:' + sc.package in self._packages:
+                self._availables[sc.__name__.lower()] = sc
+
+        self._log.info('Supported workloads available on target:')
+        self._log.info('  %s', ', '.join(self._availables.keys()))
+
+    def getInstance(self, name):
```

Comment:
Well, that's the most logical way and that it's how it was originally... but I had quite a lot of issue with the subclasses under workloads being properly identified and registered.

In `_check_availables()` there is that loop on `Workload.__subclasses__()`... AFAIR I had quite wired behaviours where it was not returning subclasses when executed from a test if this was a static method... while was working if executed from an interactive console and/or a notebook.

However, I agree that that should be better a static one... lemme address all the other comments and then I'll give it one more try before refreshing the PR.